src/languageSupport/languages/flux/lsp/connection.ts

```diff
@@ -91,6 +111,193 @@ class LspConnectionManager {
+          this._toggleSchemaSync()
+          return
+        }
+      })
+    })
+  }
+
+  _setCompositionBlockSize() {
+    const query = this._model.getValue()
+    const endLine =
+      query.split('\n').findIndex(line => line.includes(COMPOSITION_YIELD)) + 1
+    this._callbackSetSession({
+      ...this._session,
+      composition: {
+        ...this._session.composition,
+        block: {startLine: COMPOSITION_INIT_LINE, endLine},
+      },
+    })
+  }
+
+  _initComposition(schema: SchemaSelection) {
+    if (!this._session.composition.canInit) {
+      return
+    }
+
+    // msg to LSP
+    const {bucket, measurement} = schema
+    const payload = {bucket: bucket?.name}
+    if (measurement) {
+      payload['measurement'] = measurement
+    }
+    // TODO: finish LSP update first
+    // this.inject(ExecuteCommand.CompositionInit, payload)
+
+    // handlers to trigger end composition
+    this._setCompositionSyncToggle()
+    this._setCompositionIrreversibleExit(schema)
+
+    // handlers for composition block size
+    // eventually, this could be from the LSP response. onLspCompositionChange()
+    this._setCompositionBlockSize()
+    this._model.onDidChangeContent(
+      () => this._session.composition.synced && this._setCompositionBlockSize()
+    )
+
+    // set init style. (will be needed on load from persisted session.)
+    this._setCompositionBlockStyle(schema)
+  }
+
+  _endComposition(schema: SchemaSelection) {
+    // msg to LSP
+    // TODO: is there any msg to the LSP on composition end?
+    // can we have it change the yield unique identifier?
+    // this.inject(ExecuteCommand.CompositionInit, {})
+
+    // msg to monaco-editor
+    this._setCompositionBlockStyle(schema)
+  }
+
+  // hack for now. To trigger for example video.
+  private _turnOnWithFirstCall = true
+
+  onSchemaSessionChange(schema: SchemaSelection, sessionCb) {
+    this._callbackSetSession = sessionCb
+    const previousState = this._session
+    this._session = {...schema}
+
+    if (!previousState?.composition.synced && schema.composition.synced) {
+      return this._initComposition(schema)
+    }
+    if (previousState?.composition.synced && !schema.composition.synced) {
+      return this._endComposition(schema)
+    }
+    if (!isEqual(previousState?.composition.block, schema.composition.block)) {
+      return this._setCompositionBlockStyle(schema)
+    }
+    // TODO: all other updates
+    // will use this.inject() --> send to LSP
+
+    // hack for now. To trigger for example video.
+    if (this._turnOnWithFirstCall && !schema.composition.synced) {
+      this._turnOnWithFirstCall = false
+      setTimeout(() => {
+        console.error('MOCK SCHEMA INJECTION FROM USER')
+        this._toggleSchemaSync()
+      }, 5000)
+    }
+  }
+
+  onLspCompositionChange(_jsonrpcMiddlewareResponse: unknown) {
+    // TODO: middleware to read LSP response => call this method
+    // TODO: wait for successful LSP roundtrip, then use response for:
```

Comment:
`onLspCompositionChange()` = this is the method, which will be subscribing to all LSP outbound communications (after the middleware is in place --> Q4 goal).